xbmc/windowing/wayland/WinSystemWayland.cpp

```diff
@@ -466,89 +553,207 @@ bool CWinSystemWayland::SetFullScreen(bool fullScreen, RESOLUTION_INFO& res, boo
+        // each time the mouse moves without any throttling (i.e. multiple times
+        // per rendered frame).
+        // Going through all those and applying them would waste a lot of time when
+        // we already know that the size is not final and will change again anyway.
+        skippedConfigures++;
+        lastConfigureMessage = std::move(guard);
+        break;
+      case WinSystemWaylandProtocol::OUTPUT_HOTPLUG:
+        // Maybe the output that was added was the one we should be on?
+        if (m_bFullScreen)
+        {
+          CLog::LogF(LOGDEBUG, "Output hotplug in full screen mode: Re-reading resolutions");
+          UpdateResolutions();
+          // This will call SetFullScreen(), which will match the output against
+          // the information from the resolution and call set_fullscreen on the
+          // surface if it changed.
+          g_graphicsContext.SetVideoResolution(g_graphicsContext.GetVideoResolution(), true);
+        }
+        break;
+      case WinSystemWaylandProtocol::BUFFER_SCALE:
+        // Never update buffer scale if not possible to set it
+        if (m_surface.can_set_buffer_scale())
+        {
+          newScale = (reinterpret_cast<WinSystemWaylandProtocol::MsgBufferScale*> (message->data))->scale;
+        }
+        break;
+    }
   }
-  // configure is acked when the Kodi surface has actually been reconfigured
+
+  if (lastConfigureMessage)
+  {
+    if (skippedConfigures > 0)
+    {
+      CLog::LogF(LOGDEBUG, "Skipped %d configures", skippedConfigures);
+    }
+    // Wayland will tell us here the size of the surface that was actually created,
+    // which might be different from what we expected e.g. when fullscreening
+    // on an output we chose - the compositor might have decided to use a different
+    // output for example
+    // It is very important that the EGL native module and the rendering system use the
+    // Wayland-announced size for rendering or corrupted graphics output will result.
+    auto configure = reinterpret_cast<WinSystemWaylandProtocol::MsgConfigure*> (lastConfigureMessage.Get()->data);
+    CLog::LogF(LOGDEBUG, "Configure serial %u: size %dx%d state %s", configure->serial, configure->surfaceSize.Width(), configure->surfaceSize.Height(), IShellSurface::StateToString(configure->state).c_str());
+    ApplyShellSurfaceState(configure->state);
+
+    CSizeInt size = configure->surfaceSize;
+    bool sizeIncludesDecoration = true;
+
+    if (size.IsZero())
+    {
+      if (configure->state.test(IShellSurface::STATE_FULLSCREEN))
+      {
+        // Do not change current size - UpdateWithConfiguredSize must be called regardless in case
+        // scale or something else changed
+        size = m_configuredSize;
+      }
+      else
+      {
+        // Compositor has no preference and we're windowed
+        // -> adopt windowed size that Kodi wants
+        auto const& windowed = CDisplaySettings::GetInstance().GetResolutionInfo(RES_WINDOW);
```

Comment:
That is *not possible* with Wayland. The compositor has control over the width and height in a number of cases (e.g. full screen and maximized window). If the application does not follow the size that the compositor wants, it will be terminated.